flocker/provision/_digitalocean.py

```diff
@@ -0,0 +1,224 @@
+                time.sleep(1)
+                continue
+            raise
+        else:
+            return result
+
+
+def set_latest_droplet_kernel(droplet, kernel_prefix='Fedora 20 x64'):
+    """
+    Change the kernel of the droplet with ``droplet_id`` to the latest kernel
+    version with the given ``kernel_prefix``.
+    """
+    matching_kernels = [kernel for kernel in droplet.get_available_kernels()
+                        if kernel.name.startswith(kernel_prefix)]
+    latest_kernel = sorted(
+        matching_kernels,
+        key=lambda kernel: kernel.version.split('.'),
+        reverse=True)[0]
+
+    retry_if_pending(droplet.change_kernel, latest_kernel.id)
+    return latest_kernel
+
+
+def provision_digitalocean(node, package_source, distribution, token):
+    """
+    Provision flocker on this node.
+    """
+    # DO doesn't support booting the droplet's own kernel.
+    # * http://digitalocean.uservoice.com/forums/136585-digitalocean/suggestions/2814988-give-option-to-use-the-droplet-s-own-bootloader # noqa
+    # So rather than upgrade, we'll need to have new task to install the kernel
+    # package (and headers) for the DO supported kernel.
+    # The Fedora droplet default is to use a kernel that's too old for our
+    # purposes.
+    # Our documentation describes how to select a newer (DO supported) kernel
+    # for this droplet.
+    # Unfortunately this operation is only supported in the DO v2 API.
+    # * http://digitalocean.uservoice.com/forums/136585-digitalocean/suggestions/5618546-add-the-ability-to-change-kernel-via-api # noqa
+    # * https://developers.digitalocean.com/#change-the-kernel
+    # But libcloud only supports the DO v1 API
+    # * https://www.digitalocean.com/community/questions/does-libcloud-work-with-digitalocean-s-v2-api # noqa
+    # * https://issues.apache.org/jira/browse/JCLOUDS-613
+    v2client = pyocean.DigitalOcean(access_token=token)
+    v2droplet = v2client.droplet.get(node._node.id)
+
+    kernel = set_latest_droplet_kernel(v2droplet)
+    version, distribution, architecture = kernel.version.rsplit('.', 2)
+    version, release = version.split('-', 1)
+    run(
+        username='root',
+        address=node.address,
+        commands=task_install_kernel(version=version, release=release,
+                                     distribution='fc20',
+                                     architecture='x86_64')
+    )
+
+    # libcloud doesn't support shutting down DO vms.
+    # XXX Create an issue.
+    retry_if_pending(v2droplet.shutdown)
+
+    # Sorry about this, but shutdown returns the following, indicating that the
+    # droplet has halted, but it still seems to require some time before
+    # powering on.
+    # {u'status': u'completed', u'resource_id': 3797602, u'region': u'ams3', u'completed_at': u'2015-01-15T20:52:36Z', u'started_at': u'2015-01-15T20:52:31Z', u'type': u'shutdown', u'id': 41364967, u'resource_type': u'droplet'} # noqa
+    time.sleep(30)
```

Comment:
Reminder to self: Check if this can be replaced with a `while` loop waiting for something.